src/drivers/storport/libraries/StorportDrivers.qll

```diff
@@ -0,0 +1,246 @@
+/**
+ * This QL library defines classes and predicates for analyzing NDIS drivers.
+ * It provides definitions for NDIS dispatch routines, callback routines, and role types.
+ * The library also includes a typedef for the standard NDIS callback routines.
+ */
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+import cpp
+import drivers.libraries.SAL
+
+
+/** Determines if a given assignment, recursively, has a NDIS callback routine as the right-hand side. */
+private predicate isCallbackRoutineAssignment(AssignExpr ae) {
+  exists(FunctionAccess fa |
+    ae.getRValue() = fa and
+    fa.getTarget() instanceof StorportCallbackRoutine
+  )
+  or
+  ae.getRValue() instanceof AssignExpr and
+  isCallbackRoutineAssignment(ae.getRValue().(AssignExpr))
+}
+
+/** A typedef for Role Types */
+class StorportRoleTypeType extends TypedefType {
+  StorportRoleTypeType() {
+    (
+      this.getName().matches("sp_DRIVER_INITIALIZE") or
+      this.getName().matches("HW_INITIALIZE") or
+      this.getName().matches("HW_BUILDIO") or
+      this.getName().matches("HW_STARTIO") or
+      this.getName().matches("HW_INTERRUPT") or
+      this.getName().matches("HW_TIMER") or
+      this.getName().matches("HW_FIND_ADAPTER") or
+      this.getName().matches("HW_RESET_BUS") or
+      this.getName().matches("HW_ADAPTER_CONTROL") or
+      this.getName().matches("HW_PASSIVE_INITIALIZE_ROUTINE") or
+      this.getName().matches("HW_DPC_ROUTINE") or
+      this.getName().matches("HW_FREE_ADAPTER_RESOURCES") or
+      this.getName().matches("HW_PROCESS_SERVICE_REQUEST") or
+      this.getName().matches("HW_COMPLETE_SERVICE_IRP") or
+      this.getName().matches("HW_INITIALIZE_TRACING") or
+      this.getName().matches("HW_CLEANUP_TRACING") or
+      this.getName().matches("VIRTUAL_HW_FIND_ADAPTER") or
+      this.getName().matches("HW_MESSAGE_SIGNALED_INTERRUPT_ROUTINE")
+    )
+  }
+}
+
+/** A typedef for the standard NDIS callback routines. Aka Role Types */
+class StorportCallbackRoutineTypedef extends StorportRoleTypeType {
+  StorportCallbackRoutineTypedef() { this.getFile().getBaseName().matches("storport.h") }
+}
+
+/**
+ * Represents a function implementing a NDIS callback routine.
+ * Defines a function to be a callback routine iff it has a typedef
+ * in its definition which matches the NDIS callback typedefs, and it
+ * is in a NDIS driver (includes wdm.h.)

```

Comment:
```suggestion
 * is in a Storport driver (includes wdm.h.)

```

Reply:
I think "includes wdm.h" here should be "includes Storport.h", correct? Or am I misreading the comment?